File: C_Sharp/Treap/ParentTreapNode.cs
using System;

namespace Treap
{
    [Serializable]
    public class ParentTreapNode<T> : BaseTreapNode<T>
    {
        private ParentTreapNode<T> parent;

        public ParentTreapNode(T value) : base(value)
        {
        }

        protected override void LeftSetter(BaseTreapNode<T> node)
        {
            base.LeftSetter(node);
            ParentTreapNode<T> treapNode = node as ParentTreapNode<T>;
            if (treapNode != null)
            {
                treapNode.Parent = this;
            }
        }

        protected override void RightSetter(BaseTreapNode<T> node)
        {
            base.RightSetter(node);
            ParentTreapNode<T> treapNode = node as ParentTreapNode<T>;
            if (treapNode != null)
            {
                treapNode.Parent = this;
            }
        }

        public ParentTreapNode<T> Parent
        {
            get { return parent; }
            set
            {
                if (parent != null && this.Value.Equals(parent.Value))
                {
                    
                }
                parent = value;
            }
        }

        public override BaseTreapNode<T> RotateLeft()
        {
            ParentTreapNode<T> parent1 = parent;

            BaseTreapNode<T> node = base.RotateLeft();
            ((ParentTreapNode<T>) node).Parent = parent1;
            return node;
        }

        public override BaseTreapNode<T> RotateRight()
        {
            ParentTreapNode<T> parent1 = parent;
            BaseTreapNode<T> node = base.RotateRight();
            ((ParentTreapNode<T>)node).Parent = parent1;
            return node;
        }
    }
}

File: cpp/sa/ParentTreapNode.cpp
#include <stddef.h>
#include "BaseTreapNode.h"
#include "ParentTreapNode.h"

ParentTreapNode::ParentTreapNode(int value) : BaseTreapNode(value)
{
  parent = NULL;
}

void ParentTreapNode::set_left(BaseTreapNode* node)
{
  BaseTreapNode::set_left(node);	
  
  if (node != NULL)
    {
      ((ParentTreapNode *) node)->set_parent(this);
    }
}

void ParentTreapNode::set_right(BaseTreapNode* node)
{
  BaseTreapNode::set_right(node);	

  if (node != NULL)
    {
      ((ParentTreapNode *) node)->set_parent(this);
    }
}

void ParentTreapNode::set_parent(ParentTreapNode* node)
{
  parent = node;
}

ParentTreapNode* ParentTreapNode::get_parent()
{
  return parent;
}

BaseTreapNode* ParentTreapNode::rotateLeft()
{
  ParentTreapNode* parent1 = parent;
  BaseTreapNode* node = BaseTreapNode::rotateLeft();
  ((ParentTreapNode*) node)->set_parent(parent1);
  return node;
}

BaseTreapNode* ParentTreapNode::rotateRight()
{
  ParentTreapNode* parent1 = parent;
  BaseTreapNode* node = BaseTreapNode::rotateRight();
  ((ParentTreapNode*) node)->set_parent(parent1);
  return node;
}
